File: yurt-convention-api/src/main/java/com/github/liulus/yurt/convention/exception/ServiceRuntimeException.java
package com.github.liulus.yurt.convention.exception;

import com.github.liulus.yurt.convention.data.CommonCode;
import com.github.liulus.yurt.convention.data.ServiceCode;

/**
 * @author liulu
 * @version V1.0
 * @since 2020/11/9
 */
abstract class ServiceRuntimeException extends RuntimeException {

    private String code;

    protected ServiceRuntimeException() {
        super();
        this.code = CommonCode.UNKNOWN_ERROR.code();
    }

    protected ServiceRuntimeException(String message) {
        super(message);
        this.code = CommonCode.UNKNOWN_ERROR.code();
    }

    protected ServiceRuntimeException(String message, Throwable cause) {
        super(message, cause);
        this.code = CommonCode.UNKNOWN_ERROR.code();
    }

    protected ServiceRuntimeException(String code, String message) {
        super(message);
        this.code = code;
    }

    protected ServiceRuntimeException(String code, String message, Throwable cause) {
        super(message, cause);
        this.code = code;
    }

    protected ServiceRuntimeException(ServiceCode serviceCode) {
        super(serviceCode.text());
        this.code = serviceCode.code();
    }

    protected ServiceRuntimeException(ServiceCode serviceCode, Throwable cause) {
        super(serviceCode.text(), cause);
        this.code = serviceCode.code();
    }

    public String getCode() {
        return code;
    }
}

File: yurt-convention-api/src/main/kotlin/com/github/liulus/yurt/convention/exception/ServiceRuntimeException.kt
package com.github.liulus.yurt.convention.exception

import com.github.liulus.yurt.convention.enum.ErrorCode

/**
 *
 * @author liulu
 * @version V1.0
 * @since 2020/10/31
 */
abstract class ServiceRuntimeException : RuntimeException {

    var code: String = ErrorCode.UNKNOWN_ERROR.code()

    constructor() : super()

    constructor(message: String) : super(message)

    constructor(message: String, cause: Throwable) : super(message, cause) {
        if (cause is ServiceException) {
            code = cause.code
        }
    }

    constructor(cause: Throwable) : super(cause) {
        if (cause is ServiceException) {
            code = cause.code
        }
    }

    constructor(code: String, message: String) : super(message) {
        this.code = code
    }

    constructor(code: String, message: String, cause: Throwable) : super(message, cause) {
        this.code = code
    }


}


